### ISSUE TITLE
Thoughts on Mixins

### ISSUE THREAD

**Issue by arggh** [ManuelDeLeon/viewmodel#304]
I was hoping to get guidance or comments regarding ViewModel Mixins.

I've been using them on viewmodels for many things, one of them being: initializing third party libraries, when multiple components share a common pattern that uses the library.

Usually I end up having a property or few that the mixin relies on, such as "enableFunctionalityX" or "referenceToElement". What annoys me, is that there's no clear way of indicating that these properties have a meaning, when looking at code in the actual viewmodel using the mixin. The link between them becomes quite brittle.

So, I was wondering, if you had a solution in mind or any ideas? I was thinking there could be a way to specify arguments for the mixin (just the first method that I came up with):

```
mixin: {
  functionalityX: {
    name: 'funcX',
    props: { enabled: this.enableMixin(), elementToModify: this.viewEl }
   }
}
```

I'm just trying to come up with a better way of providing the mixins what they need to operate in a way that's clearly understandable both from the mixin's and viewmodel's perspective. What do you think?

Then, there's also this kind of thinking: https://medium.com/@dan_abramov/mixins-are-dead-long-live-higher-order-components-94a0d2f9e750

**Comment by ManuelDeLeon** [comment]
I wrote this fairly generic article for you: https://medium.com/@manueldeleon_94284/nothing-wrong-with-mixins-9dc40750b05d

In the case of VM, try to use mixin/share as service providers and scope them to their own variables. I only apply a mixin to the object (no scoping) when it's very simple and not much going on. I don't think I ever use it to define properties that multiple components may have. In those cases I create the properties in the component themselves.

I hope it helps.

**Comment by arggh** [comment]
Here's the key points of how this happened:

I had a component that displayed some text. I wanted the text to fit always in one line, with any length or parent dimensions, so I applied some size calculation logic to tweak the font size appropriately. Then, I wanted the same behavior on a few other components as well, so I refactored the code into a mixin. It worked, but due to different DOM element hierarchy in each component, I had to add a prop for each viewmodel that points to the correct node that contains the text to be used in the mixin's calculation. Also, I might not always want this resizing to happen, so I included a flag-property to enable or disable the mixin. It worked, but I wasn't happy with it, which got me wondering how to improve the mixin-feature of ViewModel, when in fact it was better to just put all this font-resizing-logic into a component of it's own that does this one thing and use that to wrap the text anywhere it has to fit snuggly.
